File: integrations/aws-strands/python/examples/server/api/backend_tool_rendering.py
"""Backend Tool Rendering example for AWS Strands.

Backend ``@tool`` functions execute server-side; their results flow
through Strands' normal tool-result event into the AG-UI message stream.
The frontend renders the tool call + result for the user via the message
snapshot — no frontend execution, no agent-side AG-UI event emission.
"""
import os
from pathlib import Path
from dotenv import load_dotenv

# Suppress OpenTelemetry context warnings from Strands SDK
os.environ["OTEL_SDK_DISABLED"] = "true"
os.environ["OTEL_PYTHON_DISABLED_INSTRUMENTATIONS"] = "all"

from strands import Agent, tool
from ag_ui_strands import StrandsAgent, create_strands_app
from server.model_factory import create_model
from server.settings import cors_origins

# Load environment variables from .env file
env_path = Path(__file__).parent.parent.parent / '.env'

load_dotenv(dotenv_path=env_path)

# Create model from MODEL_PROVIDER env var (default: openai)
model = create_model()

# Define backend tools for demonstration
@tool
def render_chart(chart_type: str, data: str) -> dict:
    """
    Render a chart with backend processing capabilities.
    
    Args:
        chart_type: Type of chart (bar, line, pie, etc.)
        data: Chart data in JSON format
    
    Returns:
        Chart data for frontend rendering
    """
    return {
        "chart_type": chart_type,
        "data": data[:100],
        "status": "rendered"
    }

@tool
def get_weather(location: str) -> dict:
    """
    Get weather information for a location.
    
    Args:
        location: The location to get weather for
    
    Returns:
        Weather data with temperature, conditions, humidity, wind speed
    """
    import random

    if os.environ.get("STRANDS_DEMO_FIXED_WEATHER") == "1":
        # Stable tool input for captured event tests; keep the result fully visible.
        return {
            "temperature": 72,
            "conditions": "sunny",
            "humidity": 45,
            "wind_speed": 8,
            "feels_like": 74,
        }
    
    # Simulate different weather conditions
    conditions_list = ["sunny", "cloudy", "rainy", "clear", "partly cloudy"]
    
    return {
        "temperature": random.randint(60, 85),
        "conditions": random.choice(conditions_list),
        "humidity": random.randint(30, 80),
        "wind_speed": random.randint(5, 20),
        "feels_like": random.randint(58, 88)
    }

strands_agent = Agent(
    model=model,
    tools=[get_weather, render_chart],
    system_prompt="You are a helpful assistant with backend tool rendering capabilities. You can get weather information and render charts.",
)

agui_agent = StrandsAgent(
    agent=strands_agent,
    name="backend_tool_rendering",
    description="AWS Strands agent with backend tool rendering support",
)

app = create_strands_app(agui_agent, "/", origins=cors_origins())

File: integrations/aws-strands/typescript/examples/server/api/backend-tool-rendering.ts
/**
 * Backend Tool Rendering example for AWS Strands (TypeScript).
 *
 * Demonstrates backend-executed tools. Tool results reach the frontend as
 * `TOOL_CALL_RESULT` events, which a page can render into a card of its own
 * as the dojo does, provided it reads the field names below.
 *
 * The tool shapes mirror `python/examples/server/api/backend_tool_rendering.py`
 * because the dojo's WeatherCard reads `location` from the call arguments and
 * `temperature` / `conditions` / `humidity` / `wind_speed` / `feels_like` from
 * the result. Renaming either side leaves the card showing no location and
 * every number at its zero default.
 */

import { Agent, tool } from "@strands-agents/sdk";
import { z } from "zod";
import { StrandsAgent } from "@ag-ui/aws-strands";
import { createStrandsApp } from "@ag-ui/aws-strands/server";
import { createModel } from "../model-factory";
import { demoPort, listenOrExit, runIfMain } from "../run-if-main";

const CONDITIONS = ["sunny", "cloudy", "rainy", "clear", "partly cloudy"];

function randomInt(low: number, high: number): number {
  return Math.floor(Math.random() * (high - low + 1)) + low;
}

export const getWeather = tool({
  name: "get_weather",
  description: "Get weather information for a location.",
  inputSchema: z.object({
    location: z.string().describe("The location to get weather for."),
  }),
  callback: () =>
    process.env.STRANDS_DEMO_FIXED_WEATHER === "1"
      ? {
          // Stable tool input for captured event tests; keep the result fully visible.
          temperature: 72,
          conditions: "sunny",
          humidity: 45,
          wind_speed: 8,
          feels_like: 74,
        }
      : {
          temperature: randomInt(60, 85),
          conditions: CONDITIONS[randomInt(0, CONDITIONS.length - 1)],
          humidity: randomInt(30, 80),
          wind_speed: randomInt(5, 20),
          feels_like: randomInt(58, 88),
        },
});

export const renderChart = tool({
  name: "render_chart",
  description: "Render a chart with backend processing capabilities.",
  inputSchema: z.object({
    chart_type: z.string().describe("Type of chart (bar, line, pie, etc.)"),
    data: z.string().describe("Chart data in JSON format"),
  }),
  callback: ({ chart_type, data }) => ({
    chart_type,
    data: data.slice(0, 100),
    status: "rendered",
  }),
});

const SYSTEM_PROMPT =
  "You are a helpful assistant with backend tool rendering capabilities. You can get weather information and render charts.";

export async function createBackendToolRenderingAgent(): Promise<StrandsAgent> {
  return new StrandsAgent({
    agent: new Agent({
      model: await createModel(),
      systemPrompt: SYSTEM_PROMPT,
      tools: [getWeather, renderChart],
    }),
    name: "backend_tool_rendering",
    description:
      "Strands agent that invokes backend tools and renders the results in the UI",
  });
}

runIfMain(import.meta.url, async () => {
  // Port first: it throws on a malformed PORT, and building the agent first
  // would surface a missing API key instead and hide the real complaint.
  const port = demoPort();
  const app = await createStrandsApp(await createBackendToolRenderingAgent(), {
    path: "/",
  });
  listenOrExit(app, "backend-tool-rendering", port);
});
